### PR TITLE
Dupe suggestions

### PR DESCRIPTION
**Pull request by TheYande**
#337

### CODE REVIEW

**Comment by RedGuy12** [comment]
Resolves #337

File: modules/suggestions/reactions.ts [hunk 1]
```diff
@@ -36,11 +37,56 @@ export async function addToDatabase(thread: AnyThreadChannel): Promise<void> {
 	const defaultEmoji = config.channels.suggestions?.defaultReactionEmoji;
 	const message = await thread.fetchStarterMessage().catch(() => void 0);
 	const count = (defaultEmoji?.id && message?.reactions.resolve(defaultEmoji.id)?.count) || 0;
+	const suggestionData = getSuggestionData(thread);
+	suggestionsDatabase.data = [...suggestionsDatabase.data, { ...suggestionData, count }];
+	const data = suggestionsDatabase.data as {
+		answer: typeof suggestionAnswers[number];
```
Comment (RedGuy12) [inline modules/suggestions/reactions.ts:43]: This should work without an `as`

File: modules/suggestions/reactions.ts [hunk 2]
```diff
@@ -36,11 +37,56 @@ export async function addToDatabase(thread: AnyThreadChannel): Promise<void> {
...
+	}[];
 
-	suggestionsDatabase.data = [
-		...suggestionsDatabase.data,
-		{ ...getSuggestionData(thread), count },
-	];
+	const dupes = await findDuplicates(suggestionData, data);
+	if (dupes.length == 0) return;
+	const links = dupes
+		.toSorted((a, b) => a.score - b.score)
```
Comment (RedGuy12) [inline modules/suggestions/reactions.ts:53]: Sort it correctly thr first time, don't just reverse it

File: modules/suggestions/reactions.ts [hunk 3]
```diff
@@ -36,11 +37,56 @@ export async function addToDatabase(thread: AnyThreadChannel): Promise<void> {
...
+		.toReversed()
+		.map((dupe) => {
+			return `<#${dupe.id}> ${dupe.answer}`;
+		});
+	thread.send(`
+	## Possible dupes found:\n${links.join("\n")}
```
Comment (RedGuy12) [inline modules/suggestions/reactions.ts:59]: await this

File: modules/suggestions/reactions.ts [hunk 4]
```diff
@@ -36,11 +37,56 @@ export async function addToDatabase(thread: AnyThreadChannel): Promise<void> {
...
+	let possibleDupes: {
+		id: Snowflake;
+		title: number | string;
+		answer: string;
+		score: number;
+	}[] = [];
+	for (const dbSuggestion of database) {
+		if (dbSuggestion.id == newSuggestion.id) continue;
+		const similarityThreshold = 0.5;
+		const similarity = stringSimilarity(`${newSuggestion.title}`, `${dbSuggestion.title}`);
```
Comment (RedGuy12) [inline modules/suggestions/reactions.ts:83]: string-similarity-js isn't even installed. Look at how Scradd Chat does this using didyoumean2.

File: modules/suggestions/reactions.ts [hunk 5]
```diff
@@ -36,11 +37,56 @@ export async function addToDatabase(thread: AnyThreadChannel): Promise<void> {
...
+	}[],
+) {
+	let possibleDupes: {
+		id: Snowflake;
+		title: number | string;
+		answer: string;
+		score: number;
+	}[] = [];
+	for (const dbSuggestion of database) {
+		if (dbSuggestion.id == newSuggestion.id) continue;
```
Comment (RedGuy12) [inline modules/suggestions/reactions.ts:81]: If you search for dupes before updating the database, you don't need this check

File: modules/suggestions/reactions.ts (hunk 1 above)
Comment (YandeMC) [reply modules/suggestions/reactions.ts:43]: ts throws a ` is 'readonly' and cannot be assigned to the mutable type `

File: modules/suggestions/reactions.ts [hunk 6]
```diff
@@ -36,11 +37,42 @@ export async function addToDatabase(thread: AnyThreadChannel): Promise<void> {
...
+	## Possible dupes found:\n${links.join("\n")}
+	`);
+	suggestionsDatabase.data = [...suggestionsDatabase.data, { ...suggestionData, count }];
+}
 
-	suggestionsDatabase.data = [
-		...suggestionsDatabase.data,
-		{ ...getSuggestionData(thread), count },
-	];
+export async function findDuplicates(
```
Comment (RedGuy12) [inline modules/suggestions/reactions.ts:59]: This function is only used once, and only has one line. Just inline it. That'll also fix the readonly error

**Comment by RedGuy12** [comment]
@TheYande I improved the duplicate finder to use matches from both `didyoumean2` and `match-sorter`, this gives the best results.

I also set it up to track suggestion categories internally. Every suggestion is categorized as Other by default, but the category syncs whenever the thread is updated or someone reacts. Currently, categories aren't displayed anywhere, but I use them to filter suggestions before searching for duplicates. For instance, now it won't display any website suggestions if the new thread is tagged as Editor. If the new thread is tagged as Other, no filtering is performed, and duplicates tagged as Other are always included. In the future, once the data has built up a bit, I may do something with it in `/top-suggestions`.

I had to refactor a fair chunk of the entire suggestions module to implement these features well. I also changed the display of the duplicate message to use an embed with suggestions partitioned by the answer.

In the future, I'd like to include a section of related archived suggestions, just as a reference, but not for now.

**Pull request merged**